hal/src/thumbv7em/clock/v2/dpll.rs

```diff
@@ -31,109 +31,109 @@ use typenum::U0;
+/// on [type-level enums] for more details on the pattern.
+///
+/// [type-level enums]: crate::typelevel#type-level-enum
+pub enum DpllId0 {}
 
-    impl Sealed for Dpll1 {}
+impl Sealed for DpllId0 {}
 
-    impl DpllNum for Dpll1 {
-        const NUM: usize = 1;
-    }
+impl DpllId for DpllId0 {
+    const NUM: usize = 0;
+}
 
-    impl GclkSourceMarker for Dpll1 {
-        const GCLK_SRC: GclkSourceEnum = GclkSourceEnum::DPLL1;
-    }
+/// Type-level variant representing the identity of DPLL1
+///
+/// This type is a member of several [type-level enums]. See the documentation
+/// on [type-level enums] for more details on the pattern.
+///
+/// [type-level enums]: crate::typelevel#type-level-enum
+pub enum DpllId1 {}
 
-    impl NotGclkInput for Dpll1 {}
+impl Sealed for DpllId1 {}
 
-    impl SourceMarker for Dpll1 {}
+impl DpllId for DpllId1 {
+    const NUM: usize = 1;
 }
 
 //==============================================================================
 // DpllSource
 //==============================================================================
 
-/// Source marker trait for [`Dpll`] sources
+/// Type-level `enum` for DPLL sources
 ///
-/// Note: This trait is used inconsistently; [`Pclk`] as a source is used
-/// directly in an API and therefore matching over this type is redundant.
-pub trait DpllSourceMarker: SourceMarker {
-    /// Associated constant provides a variant of a low level enum type from PAC
-    /// that is used during a HW register write
-    const DPLL_SRC: DpllSourceEnum;
+/// See the documentation / on [type-level enums] for more details on the
+/// pattern.
+///
+/// [type-level enums]: crate::typelevel#type-level-enum
+pub trait DpllSourceId {
+    /// Corresponding variant of [`DynDpllSourceId`]
+    const DYN: DynDpllSourceId;
 }
 
-/// This trait represents a [`Dpll`] source provider.
-///
-/// Note: This trait is used inconsistently; [`Pclk`] as a source is used
-/// directly in an API and therefore abstracting away through this trait is
-/// redundant.
-pub trait DpllSource: Source {
-    /// Associated type used in order to mark
-    /// [`Dpll<_, XoscDriven<_, T>>`]/[`Dpll<_, Xosc32kDriven<_, T>>`] type with
-    /// a proper `T`, according to what `source` was passed into the
-    /// [`Dpll::from_xosc`]/[`Dpll::from_xosc32k`] and to only allow calls into
-    /// [`Dpll::free`] with a matching `source`
-    type Type: DpllSourceMarker;
+impl DpllSourceId for XoscId0 {
+    const DYN: DynDpllSourceId = DynDpllSourceId::XOSC0;
+}
+
+impl DpllSourceId for XoscId1 {
+    const DYN: DynDpllSourceId = DynDpllSourceId::XOSC1;
+}
+
+impl DpllSourceId for Xosc32kId {
+    const DYN: DynDpllSourceId = DynDpllSourceId::XOSC32;
 }
 
 /// [`DpllSource`] subtrait that is used to distinguish between external 32kHz
 /// and non-32kHz oscillators
-pub trait DpllSourceXosc32k: DpllSource {}
+pub trait DpllSourceXosc32k: Driver {}
+
+impl<M, N> DpllSourceXosc32k for Enabled<Xosc32k<M>, N>
```

Comment:
Following the path of `Driver` idea, it should be possible to remove and replace this pattern with subhierarchy of `*Id` traits

```rust
pub trait DpllXosc32kBasedSourceId: DpllSourceId {}
pub trait DpllXoscBasedSourceId: DpllSourceId {}
```
and then implement it on corresponding `*Id` types. Then narrow down `T: DpllSourceId` clause of `Xosc32kDriven` and `XoscDriven` specialisations of `Dpll`  to `T: DpllXosc32kBasedSourceId` and `T: DpllXoscBasedSourceId` respectively. That should work I believe.

Reply:
I realized `DpllXosc32kBasedSourceId` can be just replaced with a specialisation for a concrete `Xosc32kId` type. The other one needs to stay to cover both `Xosc0Id` and `Xosc1Id`

Reply:
I haven't gotten to the `dpll` module yet. But you hint at one of the issues I have with the `osculp32k` and `xosc32k` modules. We have a single type that represents multiple clocks. Right now, the `Driver` trait expects that each type has exactly one `Source` type. But `OscUlp32k<Active1k, Active32k>` could conceivably implement `Driver` twice, with two different `Source` types. It breaks the consistency, which is why I've been trying so hard to find a different way to phrase it.